fuse_models/test/test_unicycle_2d_state_cost_function.cpp

```diff
@@ -56,85 +56,105 @@ TEST(CostFunction, evaluateCostFunction)
-  std::vector<fuse_core::MatrixXd> J_autodiff(num_parameter_blocks);
-  std::vector<double*> jacobians_autodiff(num_parameter_blocks);
-
-  for (size_t i = 0; i < num_parameter_blocks; ++i)
-  {
-    J_autodiff[i].resize(num_residuals, block_sizes[i]);
-    jacobians_autodiff[i] = J_autodiff[i].data();
-  }
-
-  EXPECT_TRUE(cost_function_autodiff.Evaluate(parameters, residuals.data(), jacobians_autodiff.data()));
-
-  const Eigen::IOFormat HeavyFmt(
-      Eigen::FullPrecision, 0, ", ", ";\n", "[", "]", "[", "]");
+  std::random_device dev;
+  std::mt19937 gen(dev());
+  std::uniform_real_distribution<> position_dist(0.0, 0.5);
+  std::uniform_real_distribution<> yaw_dist(-M_PI / 10.0, M_PI / 10.0);
 
-  for (size_t i = 0; i < num_parameter_blocks; ++i)
+  std::size_t N = 10;
+  for (std::size_t i = 0; i < N; i++)
   {
-    EXPECT_MATRIX_NEAR(J_autodiff[i], J[i], std::numeric_limits<double>::epsilon())
-      << "Autodiff Jacobian[" << i << "] =\n" << J_autodiff[i].format(HeavyFmt)
-      << "\nAnalytic Jacobian[" << i << "] =\n" << J[i].format(HeavyFmt);
+    // Randomly generate first state
+    double position1[] = { position_dist(gen), position_dist(gen) };
+    double yaw1[] = { yaw_dist(gen) };
+    double vel_linear1[] = { position_dist(gen), position_dist(gen) };
+    double vel_yaw1[] = { yaw_dist(gen) / 10.0 };
+    double acc_linear1[] = { vel_linear1[0] / dt, vel_linear1[1] / dt };
+
+    // Compute second state from first
+    double vel_linear2[2];
+    double position2[2];
+    double yaw2[1];
+    double vel_yaw2[1];
+    double acc_linear2[2];
+    fuse_models::predict(position1, yaw1, vel_linear1, vel_yaw1, acc_linear1, dt, position2, yaw2, vel_linear2,
+                         vel_yaw2, acc_linear2);
+
+    const double* parameters[] =
+    {
+      position1, yaw1, vel_linear1, vel_yaw1, acc_linear1,
+      position2, yaw2, vel_linear2, vel_yaw2, acc_linear2
+    };
+
+    fuse_core::Vector8d residuals;
+
+    const auto& block_sizes = cost_function.parameter_block_sizes();
+    const auto num_parameter_blocks = block_sizes.size();
+
+    const auto num_residuals = cost_function.num_residuals();
+
+    std::vector<fuse_core::MatrixXd> J(num_parameter_blocks);
+    std::vector<double*> jacobians(num_parameter_blocks);
+
+    for (size_t i = 0; i < num_parameter_blocks; ++i)
+    {
+      J[i].resize(num_residuals, block_sizes[i]);
+      jacobians[i] = J[i].data();
+    }
+
+    EXPECT_TRUE(cost_function.Evaluate(parameters, residuals.data(), jacobians.data()));
+
+    // We cannot use std::numeric_limits<double>::epsilon() tolerance because with the expected state2 above the
+    // residuals are not zero for position2.x = -4.389e-16 and yaw2 = -2.776e-16
+    EXPECT_MATRIX_NEAR(fuse_core::Vector8d::Zero(), residuals, 1e-13);
+
+    // Check jacobians are correct using a gradient checker
+    ceres::NumericDiffOptions numeric_diff_options;
+    #if !CERES_SUPPORTS_MANIFOLDS
```

Comment:
I think we should have the pre-processor macros unindented
```suggestion
#if !CERES_SUPPORTS_MANIFOLDS
```